File: examples/python/client_curve.py
"""
Irondomo Protocol client example. Uses the IDPClient API to hide all IDP aspects
Author: Matteo Ferrabone <matteo.ferrabone@gmail.com> 
"""
import os
import sys
import zmq.auth
from IronDomo  import IDPClient

def main():
    verbose = '-v' in sys.argv
    base_dir = os.path.dirname(__file__)
    keys_dir = os.path.join(base_dir, 'certificates')
    public_keys_dir = os.path.join(base_dir, 'public_keys')
    secret_keys_dir = os.path.join(base_dir, 'private_keys')
    client_secret_file = os.path.join(secret_keys_dir, "client.key_secret")
    client_public, client_secret = zmq.auth.load_certificate(client_secret_file)
    server_secret_file = os.path.join(secret_keys_dir, "server.key_secret")
    server_public, server_secret = zmq.auth.load_certificate(server_secret_file)
    client = IDPClient.IronDomoClient("tcp://localhost:5556", verbose, (server_public, client_public, client_secret))
    count = 0
    while count < 100000:
        request = "Hello world 1 -> {0}".format(count)
        try:
            reply = client.send(b"echo", request.encode())
            print(reply)
        except KeyboardInterrupt:
            break
        else:
            # also break on failure to reply:
            if reply is None:
                break
        count += 1
    print("%i requests/replies processed" % count)

if __name__ == '__main__':
    main()

File: C/examples/client_curve.c
//
//  Irondomo Protocol CURVE client example
//  Uses the idcli API to hide all MDP aspects
//

//  Lets us build this source without creating a library
#include "idcliapi.h"

int main (int argc, char *argv [])
{
    int verbose = (argc > 1 && streq (argv [1], "-v"));
    idcli_t *session = idcli_new ("tcp://localhost:5001", "EchoClient", verbose);

    zcert_t *c = zcert_new();

    const char *client_public = zcert_public_txt(c);
    const char *client_secret = zcert_secret_txt(c);
    const char *server_public = ".8Q^k*3E/4-Wg4()r^(4yTk2>qvZFDW?mXUyRPvr";

    idcli_setup_curve(session, client_public, client_secret, server_public);

    idcli_connect_to_broker(session);

    int count;
    for (count = 0; count < 2000; count++) {

        zmsg_t *request = zmsg_new ();
        zmsg_pushstr (request, "Hello");
        zmsg_pushstr (request, " ");
        zmsg_pushstr (request, "world");
        zmsg_t *reply = idcli_send (session, "echo", &request);
        if (reply)
            zmsg_destroy (&reply);
        else
            break;              //  Interrupt or failure
    }
    printf ("%d requests/replies processed\n", count);
    idcli_destroy (&session);
    return 0;
}
